File: python-examples/network-interception/api/api-interceptor.py
from playwright.async_api import Page, BrowserContext, Response
from typing import List, Callable, Awaitable, Optional
from intuned_browser import go_to_url, wait_for_network_settled
from utils.types_and_schemas import Params, Consultation


# Global storage
api_data: List[Consultation] = []
csrf_token: Optional[str] = None


def create_response_handler(pattern: str) -> Callable[[Response], Awaitable[None]]:
    async def handle_response(response: Response) -> None:
        global api_data, csrf_token

        if pattern not in response.url:
            return

        # Capture CSRF token from response headers
        if csrf_token is None:
            headers = response.headers
            csrf_token = (
                headers.get("x-csrf-token")
                or headers.get("x-xsrf-token")
                or headers.get("csrf-token")
            )

        # Capture API data
        try:
            data = await response.json()
            if isinstance(data, list):
                api_data.extend([Consultation(**item) for item in data])
            else:
                api_data.append(Consultation(**data))
        except Exception:
            pass

    return handle_response


async def automation(
    page: Page,
    params: Params,
    context: BrowserContext | None = None,
    **_kwargs,
):
    global api_data, csrf_token
    api_data = []
    csrf_token = None

    params = Params(**params)
    if not params.url:
        raise ValueError("url is required in params")

    url = params.url
    api_pattern = params.api_pattern
    max_pages = params.max_pages

    response_handler = create_response_handler(api_pattern)
    page.on("response", response_handler)

    try:
        await wait_for_network_settled(
            page=page,
            func=lambda: go_to_url(page, url),
            timeout_s=20,
        )

        current_page = 1

        while current_page < max_pages:
            next_button = await page.query_selector("#next-page-btn")
            if not next_button or await next_button.is_disabled():
                break

            items_before = len(api_data)

            await wait_for_network_settled(
                page=page,
                func=lambda: next_button.click(),
                timeout_s=20,
            )

            if len(api_data) == items_before:
                break

            current_page += 1

        return {
            "data": api_data,
            "csrf_token": csrf_token,
        }

    finally:
        page.remove_listener("response", response_handler)

File: typescript-examples/network-interception/api/api-interceptor.ts
import { Page, BrowserContext, Response } from "playwright";
import { goToUrl, withNetworkSettledWait } from "@intuned/browser";
import { z } from "zod";
import {
  consultationSchema,
  apiInterceptorParamsSchema,
} from "../utils/typesAndSchemas.js";

type Params = z.infer<typeof apiInterceptorParamsSchema>;

type Consultation = z.infer<typeof consultationSchema>;

// Global variable to store captured API data
let apiData: Consultation[] = [];

/**
 * Creates a response handler that captures responses matching the pattern.
 * Replace the pattern with the appropriate one for your API.
 */
function createResponseHandler(
  pattern: string
): (response: Response) => Promise<void> {
  return async (response: Response): Promise<void> => {
    if (response.url().includes(pattern)) {
      try {
        const data = await response.json();
        if (Array.isArray(data)) {
          apiData.push(...data);
        } else {
          apiData.push(data);
        }
      } catch {
        // Response might not be JSON
      }
    }
  };
}

/**
 * Intercepts network requests to capture paginated API data.
 * Clicks the "Next" button to load more pages and captures API responses.
 *
 * Example params:
{
    "url": "https://sandbox.intuned.dev/consultations/list",
    "api_pattern": "/rest/v1/consultations",
    "max_pages": 3
}
 *
 * This function:
 * 1. Sets up a response listener for the specified pattern
 * 2. Navigates to the URL and captures initial data
 * 3. Clicks the next page button to load more data
 * 4. Returns all captured API data
 */
async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
): Promise<Consultation[]> {
  apiData = [];

  if (!params || !params.url) {
    throw new Error("url is required in params");
  }

  const url = params.url;
  const apiPattern = params.api_pattern || "/api/";
  const maxPages = params.max_pages || 10;

  console.log(`Navigating to: ${url}`);
  console.log(`Listening for API pattern: ${apiPattern}`);
  console.log(`Max pages: ${maxPages}`);

  // Create and attach the response handler
  const responseHandler = createResponseHandler(apiPattern);
  page.on("response", responseHandler);

  try {
    // Navigate and wait until the network is settled (initial page load)
    await withNetworkSettledWait(
      async (page) => {
        await goToUrl({ page, url });
      },
      {
        page,
        timeoutInMs: 20000,
      }
    );

    console.log(`Initial load captured ${apiData.length} items`);

    // Pagination: click next button to load more pages
    let currentPage = 1;

    while (currentPage < maxPages) {
      // Check if next button exists and is visible
      // Replace "#next-page-btn" with the appropriate selector for your store
      const nextButton = await page.$("#next-page-btn");

      if (!nextButton) {
        console.log("Next button not found, stopping pagination");
        break;
      }

      const isDisabled = await nextButton.isDisabled();
      if (isDisabled) {
        console.log("Next button is disabled, reached end of data");
        break;
      }

      const itemsBefore = apiData.length;

      // Click next button and wait for network to settle
      await withNetworkSettledWait(
        async () => {
          await nextButton.click();
        },
        {
          page,
          timeoutInMs: 20000,
        }
      );

      const itemsAfter = apiData.length;
      const newItems = itemsAfter - itemsBefore;

      console.log(`Page ${currentPage + 1}: captured ${newItems} new items`);

      // If no new items were captured, stop pagination
      if (newItems === 0) {
        console.log("No new items captured, stopping pagination");
        break;
      }

      currentPage++;
    }

    console.log(`Total items captured: ${apiData.length}`);
    return apiData;
  } finally {
    // Clean up the event listener
    page.removeListener("response", responseHandler);
  }
}

export default handler;
